# Análisis de E-Commerce
Importar las librerias que se utilizarán

In [1]:
import pandas as pd
import numpy as np

Funciones que utilizaremos más adelante

In [2]:
def get_percentage(value, total):
    return value * 100 / total


Cargar los datasets

In [3]:
df_customers = pd.read_csv("olist_customers_dataset.csv")
df_geolocation = pd.read_csv("olist_geolocation_dataset.csv")
df_order_items = pd.read_csv("olist_order_items_dataset.csv")
df_order_payments = pd.read_csv("olist_order_payments_dataset.csv")
df_order_reviews = pd.read_csv("olist_order_reviews_dataset.csv")
df_orders = pd.read_csv("olist_orders_dataset.csv")
df_products = pd.read_csv("olist_products_dataset.csv")
df_sellers = pd.read_csv("olist_sellers_dataset.csv")

Analizar si alguna columna tiene un nombre poco descriptivo

No, todo chido

In [4]:
print("Customers")
print(df_customers.columns)
print("-------------------------")
print("Geolocation")
print(df_geolocation.columns)
print("-------------------------")
print("Order Items")
print(df_order_items.columns)
print("-------------------------")
print("Order Payments")
print(df_order_payments.columns)
print("-------------------------")
print("Order Reviews")
print(df_order_reviews.columns)
print("-------------------------")
print("Orders")
print(df_orders.columns)
print("-------------------------")
print("Products")
print(df_products.columns)
print("-------------------------")
print("Sellers")
print(df_sellers.columns)

Customers
Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='object')
-------------------------
Geolocation
Index(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state'],
      dtype='object')
-------------------------
Order Items
Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='object')
-------------------------
Order Payments
Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='object')
-------------------------
Order Reviews
Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='object')
-------------------------
Orders
Index(['order_id', 'customer_id', 'order_status', 'order_pur

Analizar si alguna columna tiene un tipo de dato que no le corresponda

No, todo chido

In [5]:
print("Customers")
print(df_customers.dtypes)
print("-------------------------")
print("Geolocation")
print(df_geolocation.dtypes)
print("-------------------------")
print("Order Items")
print(df_order_items.dtypes)
print("-------------------------")
print("Order Payments")
print(df_order_payments.dtypes)
print("-------------------------")
print("Order Reviews")
print(df_order_reviews.dtypes)
print("-------------------------")
print("Orders")
print(df_orders.dtypes)
print("-------------------------")
print("Products")
print(df_products.dtypes)
print("-------------------------")
print("Sellers")
print(df_sellers.dtypes)

Customers
customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object
-------------------------
Geolocation
geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                object
geolocation_state               object
dtype: object
-------------------------
Order Items
order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object
-------------------------
Order Payments
order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64
dtype: object
-------------------------
Order Reviews
review_id                  obj

Definir tipo de date Datetime a las columnas correspondientes

In [6]:
df_order_items["shipping_limit_date"] = df_order_items["shipping_limit_date"].astype('datetime64[ns]')
df_order_reviews["review_creation_date"] = df_order_reviews["review_creation_date"].astype('datetime64[ns]')
df_order_reviews["review_answer_timestamp"] = df_order_reviews["review_answer_timestamp"].astype('datetime64[ns]')
df_orders["order_purchase_timestamp"] = df_orders["order_purchase_timestamp"].astype('datetime64[ns]')
df_orders["order_approved_at"] = df_orders["order_approved_at"].astype('datetime64[ns]')
df_orders["order_delivered_carrier_date"] = df_orders["order_delivered_carrier_date"].astype('datetime64[ns]')
df_orders["order_delivered_customer_date"] = df_orders["order_delivered_customer_date"].astype('datetime64[ns]')
df_orders["order_estimated_delivery_date"] = df_orders["order_estimated_delivery_date"].astype('datetime64[ns]')
print("Order Items")
print(df_order_items.dtypes)
print("-------------------------")
print("Order Reviews")
print(df_order_reviews.dtypes)
print("-------------------------")
print("Orders")
print(df_orders.dtypes)

Order Items
order_id                       object
order_item_id                   int64
product_id                     object
seller_id                      object
shipping_limit_date    datetime64[ns]
price                         float64
freight_value                 float64
dtype: object
-------------------------
Order Reviews
review_id                          object
order_id                           object
review_score                        int64
review_comment_title               object
review_comment_message             object
review_creation_date       datetime64[ns]
review_answer_timestamp    datetime64[ns]
dtype: object
-------------------------
Orders
order_id                                 object
customer_id                              object
order_status                             object
order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    dateti

Analizar si hay columnas que tengan datos nulos

In [7]:
print("Customers")
print(df_customers.isna().sum())
print("-------------------------")
print("Geolocation")
print(df_geolocation.isna().sum())
print("-------------------------")
print("Order Items")
print(df_order_items.isna().sum())
print("-------------------------")
print("Order Payments")
print(df_order_payments.isna().sum())
print("-------------------------")
print("Order Reviews")
print(df_order_reviews.isna().sum())    # Tiene NaN
print("-------------------------")
print("Orders")
print(df_orders.isna().sum())           # Tiene NaN
print("-------------------------")
print("Products")
print(df_products.isna().sum())         # Tiene NaN
print("-------------------------")
print("Sellers")
print(df_sellers.isna().sum())

Customers
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64
-------------------------
Geolocation
geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64
-------------------------
Order Items
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64
-------------------------
Order Payments
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64
-------------------------
Order Reviews
review_id                      0
order_id                       0
review_score                   0
review_comment_title       88285
review_comment_message     58

### Datos nulos en "Orders"

Dada la naturaleza de las ventas en línea pueden suceder algunas situaciones fuera de lo esperado, por ejemplo:
- Los clientes pueden llegar a cancelar algunas compras
- Los pedidos pueden perderse en el trayecto
- Los pedidos pueden llegar a la dirección equivocada
- El comprador no recoge el paquete en la paquetería

Es comprensible que haya datos nulos aspectos como la aprobación de la order, la fecha de envío y la fecha de entrega

### Datos nulos en "Order Reviews"

Debido a que los comentarios no son obligatorios al recibir un producto, es entendible que haya valores nulos en las reviews, sin embargo, parece que es obligatorio dejar una calificación

Para facilitar el análisis exploratorio de los datos se decidió unir los datasets relacionados con las órdenes de compra/venta

In [8]:
print("Orders", df_orders.shape)
print(df_orders.columns)
print("-------------------------")
print("Order payments", df_order_payments.shape)
print(df_order_payments.columns)
print("-------------------------")
print("Order reviews", df_order_reviews.shape)
print(df_order_reviews.columns)
print("-------------------------")
print("Order items", df_order_items.shape)
print(df_order_items.columns)

Orders (99441, 8)
Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')
-------------------------
Order payments (103886, 5)
Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='object')
-------------------------
Order reviews (100000, 7)
Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='object')
-------------------------
Order items (112650, 7)
Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='object')


Droppear los duplicados (por "order_id") de order_payments y de order_reviews para definir "order_id" como index

In [9]:
# Falta agregar el dataset de items
df_orders_complete = df_orders.set_index("order_id")
df_order_payments_dropped = df_order_payments.drop_duplicates(subset=["order_id"]).set_index("order_id")
df_order_reviews_dropped = df_order_reviews.drop_duplicates(subset=["order_id"]).set_index("order_id")

Unir los dataframes para crear uno con todas las columnas

In [10]:
df_orders_complete = df_orders_complete.join(df_order_payments_dropped)
df_orders_complete = df_orders_complete.join(df_order_reviews_dropped)

In [11]:
df_orders_complete

customer_id  \
order_id                                                             
e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                                            ...   
9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

                                 order_status order_purchase_timestamp  \
order_id                                                                 
e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
53cdb2fc8bc7dce0b6741e2150273451    delivered      2018-07-24 20:41:37   
47770eb9100c2d0c44946d9cf07ec65d    delivered      2018-08-08 08:38:49   
949d5b44dbf5de918fe9c16f97b45f8a    delivered      2017-11-18 19:28:06   
ad21c59c0840e6cb83a9ceb5573f8159    delivered      2018-02-13 21:18:39   
...                                       ...                      ...   
9c5dedf39a927c1b2549525ed64a053c    delivered      2017-03-09 09:54:05   
63943bddc261676b46f01ca7ac2f7bd8    delivered      2018-02-06 12:58:58   
83c1379a015df1e13d02aae0204711ab    delivered      2017-08-27 14:46:43   
11c177c8e97725db2631073c19f07b62    delivered      2018-01-08 21:28:27   
66dea50a8b16d9b4dee7af250b4be1a5    delivered      2018-03-08 20:57:30   

                                   order_approved_at  \
order_id                                               
e481f51cbdc54678b7cc49136f2d6af7 2017-10-02 11:07:15   
53cdb2fc8bc7dce0b6741e2150273451 2018-07-26 03:24:27   
47770eb9100c2d0c44946d9cf07ec65d 2018-08-08 08:55:23   
949d5b44dbf5de918fe9c16f97b45f8a 2017-11-18 19:45:59   
ad21c59c0840e6cb83a9ceb5573f8159 2018-02-13 22:20:29   
...                                              ...   
9c5dedf39a927c1b2549525ed64a053c 2017-03-09 09:54:05   
63943bddc261676b46f01ca7ac2f7bd8 2018-02-06 13:10:37   
83c1379a015df1e13d02aae0204711ab 2017-08-27 15:04:16   
11c177c8e97725db2631073c19f07b62 2018-01-08 21:36:21   
66dea50a8b16d9b4dee7af250b4be1a5 2018-03-09 11:20:28   

                                 order_delivered_carrier_date  \
order_id                                                        
e481f51cbdc54678b7cc49136f2d6af7          2017-10-04 19:55:00   
53cdb2fc8bc7dce0b6741e2150273451          2018-07-26 14:31:00   
47770eb9100c2d0c44946d9cf07ec65d          2018-08-08 13:50:00   
949d5b44dbf5de918fe9c16f97b45f8a          2017-11-22 13:39:59   
ad21c59c0840e6cb83a9ceb5573f8159          2018-02-14 19:46:34   
...                                                       ...   
9c5dedf39a927c1b2549525ed64a053c          2017-03-10 11:18:03   
63943bddc261676b46f01ca7ac2f7bd8          2018-02-07 23:22:42   
83c1379a015df1e13d02aae0204711ab          2017-08-28 20:52:26   
11c177c8e97725db2631073c19f07b62          2018-01-12 15:35:03   
66dea50a8b16d9b4dee7af250b4be1a5          2018-03-09 22:11:59   

                                 order_delivered_customer_date  \
order_id                                                         
e481f51cbdc54678b7cc49136f2d6af7           2017-10-10 21:25:13   
53cdb2fc8bc7dce0b6741e2150273451           2018-08-07 15:27:45   
47770eb9100c2d0c44946d9cf07ec65d           2018-08-17 18:06:29   
949d5b44dbf5de918fe9c16f97b45f8a           2017-12-02 00:28:42   
ad21c59c0840e6cb83a9ceb5573f8159           2018-02-16 18:17:02   
...                                                        ...   
9c5dedf39a927c1b2549525ed64a053c           2017-03-17 15:08:01   
63943bddc261

Revisar si hay datos nulos

In [12]:
df_orders_complete.isna().sum()

customer_id                          0
order_status                         0
order_purchase_timestamp             0
order_approved_at                  160
order_delivered_carrier_date      1783
order_delivered_customer_date     2965
order_estimated_delivery_date        0
payment_sequential                   1
payment_type                         1
payment_installments                 1
payment_value                        1
review_id                            0
review_score                         0
review_comment_title             87744
review_comment_message           57886
review_creation_date                 0
review_answer_timestamp              0
dtype: int64

### Limpieza de datos nulos

In [13]:
df_orders_complete = df_orders_complete.dropna(subset=["payment_sequential", "payment_type", "payment_installments", "payment_value"])
df_orders_complete = df_orders_complete.fillna(value="")
df_orders_complete.isna().sum()

customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
payment_sequential               0
payment_type                     0
payment_installments             0
payment_value                    0
review_id                        0
review_score                     0
review_comment_title             0
review_comment_message           0
review_creation_date             0
review_answer_timestamp          0
dtype: int64

### Analizar el estado de las órdenes

In [14]:
order_status_count = df_orders_complete.groupby(["order_status"]).size().reset_index(name='count').sort_values(by="count", ascending=False)
order_status_count

order_status  count
3    delivered  96477
6      shipped   1107
1     canceled    625
7  unavailable    609
4     invoiced    314
5   processing    301
2      created      5
0     approved      2

In [15]:
percentages = get_percentage(order_status_count["count"], order_status_count["count"].sum())

print("Estado      | Porcentaje")
print("------------------------")
for index in order_status_count.index:
    print(f'{order_status_count["order_status"][index]:<12}| {np.round(percentages[index], 2)}%')

Estado      | Porcentaje
------------------------
delivered   | 97.02%
shipped     | 1.11%
canceled    | 0.63%
unavailable | 0.61%
invoiced    | 0.32%
processing  | 0.3%
created     | 0.01%
approved    | 0.0%


Analizando los estados de las órdenes se llego a la conclusión de que la mayor son entregadas con éxito, y que solo un pequeño porcentaje (menos del 1%) sufrieron algún tipo de percance

También encontramos que hay un estado de "unavailable" del cual no sabemos su significado

### Analizar el método de pago

In [16]:
order_payment_type_count = df_orders_complete.groupby(["payment_type"]).size().reset_index(name='count').sort_values(by="count", ascending=False)

In [17]:
percentages = get_percentage(order_payment_type_count["count"], order_payment_type_count["count"].sum())

print("Tipo de pago | Porcentaje")
print("-------------------------")
for index in order_payment_type_count.index:
    print(f'{order_payment_type_count["payment_type"][index]:<13}| {np.round(percentages[index], 2)}%')

Tipo de pago | Porcentaje
-------------------------
credit_card  | 75.81%
boleto       | 19.9%
voucher      | 2.75%
debit_card   | 1.54%
not_defined  | 0.0%


El método de pago "boleto" significa "billete" en español, es decir, efectivo

Analizando los métodos de pago de las órdenes se llegó a la conclusión de que es bastante la gente que utiliza tarjeta de crédito, 3/4 partes de todas las órdenes utilizan ese método

### Análisis del valor de los pagos

In [18]:
 df_orders_complete["payment_value"].describe()

count    99440.000000
mean       158.334781
std        219.518488
min          0.000000
25%         60.050000
50%        103.350000
75%        175.130000
max      13664.080000
Name: payment_value, dtype: float64

En busca de conocer el pago promedio por transacción se usó el comando "describe" para obtener los datos más relevantes. El gasto promedio por transacción es de 158 reales brasileños, lo que equivale a poco menos de 600 pesos mexicanos

Al ver esa descripción encontramos que existen pagos cuyo valor es igual a cero, así que decidimos investigar al respecto

In [19]:
 df_orders_complete[df_orders_complete["payment_value"] == 0]

customer_id  \
order_id                                                             
00b1cb0320190ca0daa2c88b35206009  3532ba38a3fd242259a514ac2b6ae6b6   
4637ca194b6387e2d538dc89b124b0ee  a73c1f73f5772cf801434bf984b0b1a7   
c8c528189310eaa44a745b8d9d26908b  197a2a6a77da93f678ea0d379f21da0a   

                                 order_status order_purchase_timestamp  \
order_id                                                                 
00b1cb0320190ca0daa2c88b35206009     canceled      2018-08-28 15:26:39   
4637ca194b6387e2d538dc89b124b0ee     canceled      2018-09-03 14:14:25   
c8c528189310eaa44a745b8d9d26908b     canceled      2018-08-28 20:05:14   

                                 order_approved_at  \
order_id                                             
00b1cb0320190ca0daa2c88b35206009                     
4637ca194b6387e2d538dc89b124b0ee                     
c8c528189310eaa44a745b8d9d26908b                     

                                 order_delivered_carrier_date  \
order_id                                                        
00b1cb0320190ca0daa2c88b35206009                                
4637ca194b6387e2d538dc89b124b0ee                                
c8c528189310eaa44a745b8d9d26908b                                

                                 order_delivered_customer_date  \
order_id                                                         
00b1cb0320190ca0daa2c88b35206009                                 
4637ca194b6387e2d538dc89b124b0ee                                 
c8c528189310eaa44a745b8d9d26908b                                 

                                 order_estimated_delivery_date  \
order_id                                                         
00b1cb0320190ca0daa2c88b35206009                    2018-09-12   
4637ca194b6387e2d538dc89b124b0ee                    2018-09-10   
c8c528189310eaa44a745b8d9d26908b                    2018-09-11   

                                  payment_sequential payment_type  \
order_id                                                            
00b1cb0320190ca0daa2c88b35206009                 1.0  not_defined   
4637ca194b6387e2d538dc89b124b0ee                 1.0  not_defined   
c8c528189310eaa44a745b8d9d26908b                 1.0  not_defined   

                                  payment_installments  payment_value  \
order_id                                                                
00b1cb0320190ca0daa2c88b35206009                   1.0            0.0   
4637ca194b6387e2d538dc89b124b0ee                   1.0            0.0   
c8c528189310eaa44a745b8d9d26908b                   1.0            0.0   

                                                         review_id  \
order_id                                                             
00b1cb0320190ca0daa2c88b35206009  9c0d840dfe562debfa52a1792bd64bff   
4637ca194b6387e2d538dc89b124b0ee  9efef4d70190c1bdb9b12872cc8144fc   
c8c528189310eaa44a745b8d9d26908b  b54d02f7f1520b3995f84b77ab0dacb3   

                                  review_score       review_comment_title  \
order_id                                                                    
00b1cb0320190ca0daa2c88b35206009             1  Entregaram produto errado   
4637ca194b6387e2d538dc89b124b0ee             3           Produto Avariado   
c8c528189310eaa44a745b8d9d26908b             1  Péssima meu produtdefeito   

                                                             review_comment_message  \
order_id                                                                              
00b1cb0320190ca0daa2c88b35206009  Comprei dois fones de ouvido com valor de R$ 5...   
4637ca194b6387e2d538dc89b124b0ee  Recebi o Diario do Bebê c um pequeno rasgo na ...   
c8c528189310eaa44a745b8d9d26908b                                                      

                                 review_creation_date review_answer_timestamp  
order_id                                                                       
00b1cb0320190ca0daa2c88b35206

Encontramos que las órdenes de los productos con valor igual a cero corresponden a casos en los que la compra fue cancelada debido a fallas en el producto o entrega de un producto equivocado